### ISSUE TITLE
JSON encoding error when enqueuing Mobile Analytics events

### ISSUE THREAD

**Issue by jmglov** [aws/aws-sdk-unity#43]
I'm using aws-sdk-unity 1.0.5 on Unity 5.0.1.f1, and Mobile Analytics is not working at all on iPhone hardware. I did some digging, and tracked the error down to [AmazonMADefaultDeliveryClient. EnqueueEventsForDelivery](https://github.com/aws/aws-sdk-unity/blob/dd54a862878f2fe7ce37c7c3e144f90e78faa77c/Assets/AWSUnitySDK/MobileAnalytics/Amazon.MobileAnalyticsManager/Delivery/AmazonMADefaultDeliveryClient.cs#L95):

```c#
        public void EnqueueEventsForDelivery(Amazon.MobileAnalytics.Model.Event eventObject)
        {
            ThreadPool.QueueUserWorkItem(new WaitCallback(delegate 
            {   
                string eventString = JsonMapper.ToJson(eventObject);
```

The `JsonMapper.ToJson()` call throws an exception:

```
[MobileAnalyticsDeliveryClient] ERROR: Exception encoding event _session.start to JSON: System.ExecutionEngineException: Attempting to JIT compile method '(wrapper delegate-invoke) System.Reflection.MonoProperty/Getter`2<Amazon.MobileAnalytics.Model.Event, System.Collections.Generic.Dictionary`2<string, string>>:invoke_Dictionary`2<string, string>__this___Event (Amazon.MobileAnalytics.Model.Event)' while running with --aot-only.

  at System.Reflection.MonoProperty.GetterAdapterFrame[Event,Dictionary`2] (System.Reflection.Getter`2 getter, System.Object obj) [0x00000] in <filename unknown>:0 
  at System.Reflection.MonoProperty.GetValue (System.Object obj, System.Object[] index) [0x00000] in <filename unknown>:0 
  at ThirdParty.Json.LitJson.JsonMapper.WriteValue (System.Object obj, ThirdParty.Json.LitJson.JsonWriter writer, Boolean writer_is_private, Int32 depth) [0x00000] in <filename unknown>:0 
  at ThirdParty.Json.LitJson.JsonMapper.ToJson (System.Object obj) [0x00000] in <filename unknown>:0 
  at Amazon.MobileAnalyticsManager.Delivery.AmazonMADefaultDeliveryClient+<EnqueueEventsForDelivery>c__AnonStorey77.<>m__96 (System.Object ) [0x00000] in <filename unknown>:0
```

I'll see if I can patch this, but if anyone has any ideas in the meantime, I'd love to hear them.

**Comment by karthiksaligrama** [comment]
There are some known issues with the sdk in unity 5 and we are working towards bringing support for unity 5. Until then we would advise to use unity 4.x

**Comment by jmglov** [comment]
Unfortunately, 4.x is not an option. I'll eagerly await your update.

Is it possible to put your develop branch on GitHub? I'd love to have a
look at how you're fixing some stuff so I can patch critical bits of the
SDK in the meantime.

Are the known Unity 5 issues listed somewhere?

On Monday, 13 April 2015, Karthik Saligrama <notifications@github.com>
wrote:

> There are some known issues with the sdk in unity 5 and we are working
> towards bringing support for unity 5. Until then we would advise to use
> unity 4.x
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aws/aws-sdk-unity/issues/43#issuecomment-92450874>.
>


-- 
Sent from my iPhone. Please excuse the brevity and potential typos.

**Comment by jmglov** [comment]
I hacked around this by using Json.Net (which our project already includes) for the problematic line.

However, Cognito authentication is broken in Unity 5 due to a Unity bug which causes **all** HTTPS requests to bomb out with a `NSURLErrorDomain error -1012`:

http://forum.unity3d.com/threads/unity-5-parse-ios-nsurlerrordomain-error-1012.308569/

Luckily, Unity claim that the fix will be out tomorrow, and in the meantime, there is a hacktastic manual work-around that involves editing `WWWConnection.mm` in the Unity-generated iOS project:

http://forum.unity3d.com/threads/unity-5-parse-ios-nsurlerrordomain-error-1012.308569/#post-2047219

Ugh.

**Comment by jmglov** [comment]
The saga continues. After applying the HTTPS hotfix as described above, now I'm getting a new error in the Cognito `GetCredentialsAsync` flow:

```
[CognitoAWSCredentials] ERROR: The requested value 'noerror' was not found.
UnityEngine.Debug:Internal_Log(Int32, String, Object)
UnityEngine.Debug:LogError(Object)
Amazon.Unity3D.AmazonLogging:LogError(String, String)
Amazon.Runtime.<InvokeHelper>c__AnonStorey12:<>m__1(AmazonServiceResult)
Amazon.Unity3D.<FireCallbackOnCoRoutine>c__Iterator2:MoveNext()
UnityEngine.MonoBehaviour:StartCoroutine_Auto(IEnumerator)
UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
Amazon.Unity3D.AmazonMainThreadDispatcher:Update()
```

I'm trying to track it down, but if any Amazonians can point me in the right direction, I'd be very grateful.

**Comment by karthiksaligrama** [comment]
I think there is a difference in the way the http status is present in the header. can you confirm that you customization of the WWWConnection.mm in iOS doesn't change the format of Http Status header?

**Comment by jmglov** [comment]
I think you're right, as I've traced the error down to `WWWResponseData.StatusCode`:

```c#
        public HttpStatusCode StatusCode
        {
            get
            {
                return (HttpStatusCode)Enum.Parse(typeof(HttpStatusCode),
                                         ResponseHeaders["STATUS"].Substring(13).Replace(" ", ""));

            }
        }
```

Checking `WWWConnection.mm` right now.

Thanks for the rapid response! We have a launch scheduled for Monday, so even a hacked-up version of MA is vital. Too bad our Android build only works in Unity 5. :(

**Comment by jmglov** [comment]
@karthiksaligrama I'm having trouble tracking down the `HttpStatusCode` enum. Is that something in the aws-unity-sdk code or a Unity thing?

**Comment by jmglov** [comment]
@karthiksaligrama : Thanks to your clue, I got things working. Here's my patch to `WWWResponseData` to work around whatever Unity is doing to the status codes:

```patch
diff --git a/Assets/AWSUnitySDK/AWSCore/Amazon.Runtime/Internal/Transform/WWWResponseData.cs b/Assets/AWSUnitySDK/AWSCore/Amazon.Runtime/Internal/Transform/WWWResponseData.cs
index 49d26fd..ca55a92 100644
--- a/Assets/AWSUnitySDK/AWSCore/Amazon.Runtime/Internal/Transform/WWWResponseData.cs
+++ b/Assets/AWSUnitySDK/AWSCore/Amazon.Runtime/Internal/Transform/WWWResponseData.cs
@@ -18,6 +18,8 @@ using System.Threading;
 
 using UnityEngine;
 
+using Amazon.Unity3D;
+
 namespace Amazon.Runtime.Internal.Transform
 {
     /// <summary>
@@ -53,6 +55,8 @@ namespace Amazon.Runtime.Internal.Transform
 
         public byte[] GetBytes()
         {
+            AmazonLogging.LogInfo("WWWResponseData","Getting bytes");
+
             return this._bytes;
         }
 
@@ -70,9 +74,13 @@ namespace Amazon.Runtime.Internal.Transform
         {
             get
             {
-                return (HttpStatusCode)Enum.Parse(typeof(HttpStatusCode),
-                                         ResponseHeaders["STATUS"].Substring(13).Replace(" ", ""));
-
+                var statusHeader = ResponseHeaders["STATUS"].Substring(13).Replace(" ", "");
+                try {
+                    return (HttpStatusCode)Enum.Parse(typeof(HttpStatusCode), statusHeader, false);
+                } catch (System.Exception exc) {
+                    AmazonLogging.LogInfo("WWWResponseData","Status header " + statusHeader + " not valid HttpStatusCode");
+                    return HttpStatusCode.OK;
+                }
             }
         }
```

It looks like the status headers that are causing the problem are `noerror` and `accepted`. There is in fact an [`System.Net.HttpStatusCode.Accepted`](https://msdn.microsoft.com/en-us/library/system.net.httpstatuscode(v=vs.80).aspx), but the case is seeming to muddle things up, even though I've told `Enum.Parse()` that the string should not be case-sensitive.

This is a terrible hack, but events are making it into MA, so I'm unstuck for now.

Unity is planning to release 5.0.1p1 on 15 April (tomorrow as of this posting), which should fix the HTTPS issue in the same way as the manual fix I applied. This means that aws-sdk-unity will almost certainly break with handling HTTP status codes, but hopefully your team has the info to fix this properly.

Thanks again for your attention to this issue!

**Comment by karthiksaligrama** [comment]
Cool. We will add this fix with other unity 5 fixes to our sdk.

**Comment by karthiksaligrama** [comment]
Fixed in 2.0.0.0
